From 585c62fe7e156768fb4ca49e32b855513ab39ef2 Mon Sep 17 00:00:00 2001
From: Masato TOYOSHIMA <phoepsilonix@phoepsilonix.love>
Date: Sun, 30 Jul 2023 09:01:05 +0900
Subject: [PATCH] In the _alpm_download called from the alpm_db_update, the
 options of the SIG file are utilized.

_alpm_download called from alpm_db_update should add a process to download sig files using payload->download_signature and payload->signature_optional.
---
 PKGBUILD                        | 12 ++++++-----
 pacman-download-signature.patch | 36 +++++++++++++++++++++++++++++++++
 2 files changed, 43 insertions(+), 5 deletions(-)
 create mode 100644 pacman-download-signature.patch

diff --git a/PKGBUILD b/PKGBUILD
index ab41e10..7bac880 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -10,7 +10,7 @@
 pkgname=pacman
 pkgver=6.0.2
 _pkgver=1.9.1
-pkgrel=13
+pkgrel=15
 pkgdesc="A library-based package manager with dependency support"
 arch=('x86_64')
 url="https://www.archlinux.org/pacman/"
@@ -45,16 +45,18 @@ source=(https://sources.archlinux.org/other/pacman/$pkgname-$pkgver.tar.xz{,.sig
         pacman-strip-include-o-files-similar-to-kernel-modules.patch::https://gitlab.archlinux.org/pacman/pacman/-/commit/de11824527ec4e2561e161ac40a5714ec943543c.patch
         pacman.conf
         makepkg.conf
+        pacman-download-signature.patch
         pacman-sync-first-option.patch)
 sha256sums=('7d8e3e8c5121aec0965df71f59bedf46052c6cf14f96365c4411ec3de0a4c1a5'
             'SKIP'
             'SKIP'
-            '522b789e442b3bb3afa7ea3fa417a99554f36ec00de3986cbe92c80f09a7db99'
-            'dab7c70fb9d77d702069bb57f5a12496b463d68ae20460fb0a3ffcb4791321a9'
-            '0b56c61eac3d9425d68faa2eccbaefdc5ed422b643974ae829eaca0460043da1'
-            'acd0b149b6324dc1eca3cd2d3b30df6ef64c5653e83523d77200ec593e01d2a7'
+            '6fed94326b9ecfbb438ab17a4576b5e9d52ecc0f4574f29d46adfde3886dee03'
+            'd2bc104788290e3de829a232590b66ad745cf5b4762a01acc1191ebf70fef114'
+            'f3d4f39ef24e312f610cbb3439fb02bc6b8829e37bcf1a50ae50cd0a69bde5d0'
+            'd87d0c9957c613fda272553bee58140349d151ae399f346ddaf6d75ee5916312'
             'fd5633e107d4082d0ce96f8afe5db08819a24d5b243413ba790ecfb864e36cd4'
             '40c8c0f874a3ce1e2290be95f79b772e1e3661b7e99608e1742349eb192c0f5a'
+            '7adeb0a172c46e3389f03c1b64479206979c7248573a0cf835fb7950fbc9dd1e'
             '8167155d3a3e15fc4a1b1e989fdb826779e7b3690a52e2ca9d307ae0b1550e1d')
 
 prepare() {
diff --git a/pacman-download-signature.patch b/pacman-download-signature.patch
new file mode 100644
index 0000000..18d6f31
--- /dev/null
+++ b/pacman-download-signature.patch
@@ -0,0 +1,36 @@
+diff --git a/lib/libalpm/dload.c b/lib/libalpm/dload.c
+index 1bf9b4d..ea92acf 100644
+--- a/lib/libalpm/dload.c
++++ b/lib/libalpm/dload.c
+@@ -964,6 +964,16 @@ int _alpm_download(alpm_handle_t *handle,
+ 
+ 			if(payload->fileurl) {
+ 				ret = handle->fetchcb(handle->fetchcb_ctx, payload->fileurl, localpath, payload->force);
++				if(payload->download_signature || payload->signature_optional) {
++					char *fileurl;
++
++					size_t len = strlen(payload->fileurl) + 4 + 2;
++					MALLOC(fileurl, len, RET_ERR(handle, ALPM_ERR_MEMORY, -1));
++					snprintf(fileurl, len, "%s.sig", payload->fileurl);
++					int ret2 = -1;
++					ret2 = handle->fetchcb(handle->fetchcb_ctx, fileurl, localpath, payload->force);
++					free(fileurl);
++				}
+ 			} else {
+ 				for(s = payload->servers; s && ret == -1; s = s->next) {
+ 					const char *server = s->data;
+@@ -975,6 +985,14 @@ int _alpm_download(alpm_handle_t *handle,
+ 
+ 					ret = handle->fetchcb(handle->fetchcb_ctx, fileurl, localpath, payload->force);
+ 					free(fileurl);
++				        if(payload->download_signature || payload->signature_optional) {
++						len = len + 4;
++						MALLOC(fileurl, len, RET_ERR(handle, ALPM_ERR_MEMORY, -1));
++						snprintf(fileurl, len, "%s/%s.sig", server, payload->filepath);
++						int ret2 = -1;
++						ret2 = handle->fetchcb(handle->fetchcb_ctx, fileurl, localpath, payload->force);
++						free(fileurl);
++					}
+ 				}
+ 			}
+ 
-- 
2.41.0

